builtin/plugins/dposv3/dpos.go

```diff
@@ -2658,3 +2606,98 @@ func Initialize(ctx contract.Context, initState *InitializationState) error {
 
 	return nil
 }
+
+func listDelegations(ctx contract.StaticContext, req *ListDelegationsRequest) (*ListDelegationsResponse, error) {
+	if req.Candidate == nil {
+		return nil, logStaticDposError(ctx, errors.New("ListDelegations called with req.Candidate == nil"), req.String())
+	}
+
+	delegations, err := loadDelegationList(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	total := common.BigZero()
+	candidateDelegations := make([]*Delegation, 0)
+	for _, d := range delegations {
+		if loom.UnmarshalAddressPB(d.Validator).Compare(loom.UnmarshalAddressPB(req.Candidate)) != 0 {
+			continue
+		}
+
+		delegation, err := GetDelegation(ctx, d.Index, *d.Validator, *d.Delegator)
+		if err == contract.ErrNotFound {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		candidateDelegations = append(candidateDelegations, delegation)
+		total = total.Add(total, &delegation.Amount.Value)
+	}
+
+	return &ListDelegationsResponse{
+		Delegations:     candidateDelegations,
+		DelegationTotal: &types.BigUInt{Value: *total},
+	}, nil
+}
+
+func listValidators(ctx contract.StaticContext, req *ListValidatorsRequest) (*ListValidatorsResponse, error) {
+	validators, err := ValidatorList(ctx)
+	if err != nil {
+		return nil, logStaticDposError(ctx, err, req.String())
+	}
+
+	chainID := ctx.Block().ChainID
+
+	displayStatistics := make([]*ValidatorStatistic, 0)
+	for _, validator := range validators {
+		address := loom.Address{ChainID: chainID, Local: loom.LocalAddressFromPublicKey(validator.PubKey)}
+
+		// get validator statistics
+		stat, _ := GetStatistic(ctx, address)
+		if stat == nil {
+			stat = &ValidatorStatistic{
+				Address: address.MarshalPB(),
+			}
+		}
+		displayStatistics = append(displayStatistics, stat)
+	}
+
+	return &ListValidatorsResponse{
+		Statistics: displayStatistics,
+	}, nil
+}
+
+func TotalStaked(ctx contract.StaticContext, bootstrapNodes map[string]bool) (*types.BigUInt, error) {
+	response, err := listValidators(ctx, &ListValidatorsRequest{})
+	if err != nil {
+		return nil, err
+	}
+	statistics := map[string]*ValidatorStatistic{}
+	for _, statistic := range response.Statistics {
+		nodeAddr := loom.UnmarshalAddressPB(statistic.Address)
+		if _, ok := bootstrapNodes[strings.ToLower(nodeAddr.String())]; !ok {
+			statistics[statistic.Address.String()] = statistic
+		}
+	}
+
+	candidates, err := LoadCandidateList(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	totalStaked := &types.BigUInt{Value: *loom.NewBigUIntFromInt(0)}
+	for _, candidate := range candidates {
+		response, err := listDelegations(ctx, &ListDelegationsRequest{Candidate: candidate.Address})
```

Comment:
Do we really need to load all delegations? We don't keep track of this amount per-validator?

Reply:
Unfortunately, we don't.

Reply:
Well, this is going to load ALL the delegations N times over, once for each candidate, should only load those once.

Reply:
This should be fine. ListDelegation is designed to be called this way. It only loads only delegations of a particular candidate, not all delegations.

Reply:
`loadDelegationList` skips loading the individual delegation records for irrelevant candidates, but it still loads a giant "index" of all the delegations, each and every time unless you use the cached storage.

Reply:
Fine, the logic was copied from `ListAllDelegations` but yeah it can be improved. Done